// azurite/dist/src/table/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALID_TABLE_AUDIENCES = exports.SECONDARY_SUFFIX = exports.TABLE_SERVICE_PERMISSION = exports.ValidAPIVersions = exports.QUERY_RESULT_MAX_NUM = exports.DEFAULT_TABLE_PERSISTENCE_ARRAY = exports.RETURN_CONTENT = exports.RETURN_NO_CONTENT = exports.ODATA_TYPE = exports.XML_METADATA = exports.FULL_METADATA_ACCEPT = exports.MINIMAL_METADATA_ACCEPT = exports.NO_METADATA_ACCEPT = exports.SUPPORTED_QUERY_OPERATOR = exports.MethodConstants = exports.HeaderConstants = exports.ENTITY_SIZE_MAX = exports.BODY_SIZE_MAX = exports.VERSION = exports.TABLE_API_VERSION = exports.DEFAULT_TABLE_CONTEXT_PATH = exports.TABLE_STATUSCODE = exports.DEFAULT_KEY_MAX_LENGTH = exports.DEFAULT_TABLE_PERSISTENCE_PATH = exports.DEFAULT_ENABLE_DEBUG_LOG = exports.DEFAULT_ENABLE_ACCESS_LOG = exports.DEFAULT_TABLE_LISTENING_PORT = exports.DEFAULT_TABLE_SERVER_HOST_NAME = exports.DEFAULT_TABLE_LOKI_DB_PATH = exports.DEFAULT_TABLE_EXTENT_LOKI_DB_PATH = void 0;
exports.DEFAULT_TABLE_EXTENT_LOKI_DB_PATH = "__azurite_db_table_extent__.json";
exports.DEFAULT_TABLE_LOKI_DB_PATH = "__azurite_db_table__.json";
exports.DEFAULT_TABLE_SERVER_HOST_NAME = "127.0.0.1"; // Change to 0.0.0.0 when needs external access
exports.DEFAULT_TABLE_LISTENING_PORT = 10002;
exports.DEFAULT_ENABLE_ACCESS_LOG = true;
exports.DEFAULT_ENABLE_DEBUG_LOG = true;
exports.DEFAULT_TABLE_PERSISTENCE_PATH = "__tablestorage__";
exports.DEFAULT_KEY_MAX_LENGTH = 512;
var TABLE_STATUSCODE;
(function (TABLE_STATUSCODE) {
    TABLE_STATUSCODE[TABLE_STATUSCODE["CREATED"] = 201] = "CREATED";
    TABLE_STATUSCODE[TABLE_STATUSCODE["NOCONTENT"] = 204] = "NOCONTENT";
})(TABLE_STATUSCODE || (exports.TABLE_STATUSCODE = TABLE_STATUSCODE = {}));
exports.DEFAULT_TABLE_CONTEXT_PATH = "azurite_table_context";
exports.TABLE_API_VERSION = "2024-05-04";
exports.VERSION = "3.30.0";
// Max Body size is 4 MB
exports.BODY_SIZE_MAX = 1024 * 1024 * 4;
// Max Entity sizxe is 1 MB
exports.ENTITY_SIZE_MAX = 1024 * 1024;
exports.HeaderConstants = {
    SERVER: "Server",
    APPLICATION_JSON: "application/json",
    AUTHORIZATION: "authorization",
    CONTENT_MD5: "content-md5",
    CONTENT_TYPE: "content-type",
    CONTENT_LENGTH: "content-length",
    DATE: "date",
    X_MS_DATE: "x-ms-date",
    X_MS_VERSION: "x-ms-version",
    ACCEPT: "accept",
    PREFER: "Prefer",
    ORIGIN: "origin",
    VARY: "Vary",
    ACCESS_CONTROL_EXPOSE_HEADERS: "Access-Control-Expose-Headers",
    ACCESS_CONTROL_ALLOW_ORIGIN: "Access-Control-Allow-Origin",
    ACCESS_CONTROL_ALLOW_CREDENTIALS: "Access-Control-Allow-Credentials",
    ACCESS_CONTROL_ALLOW_METHODS: "Access-Control-Allow-Methods",
    ACCESS_CONTROL_ALLOW_HEADERS: "Access-Control-Allow-Headers",
    ACCESS_CONTROL_MAX_AGE: "Access-Control-Max-Age",
    ACCESS_CONTROL_REQUEST_METHOD: "access-control-request-method",
    ACCESS_CONTROL_REQUEST_HEADERS: "access-control-request-headers"
};
exports.MethodConstants = {
    OPTIONS: "OPTIONS"
};
exports.SUPPORTED_QUERY_OPERATOR = ["eq", "gt", "ge", "lt", "le", "ne"];
exports.NO_METADATA_ACCEPT = "application/json;odata=nometadata";
exports.MINIMAL_METADATA_ACCEPT = "application/json;odata=minimalmetadata";
exports.FULL_METADATA_ACCEPT = "application/json;odata=fullmetadata";
exports.XML_METADATA = "application/atom+xml";
exports.ODATA_TYPE = "@odata.type";
exports.RETURN_NO_CONTENT = "return-no-content";
exports.RETURN_CONTENT = "return-content";
exports.DEFAULT_TABLE_PERSISTENCE_ARRAY = [
    {
        locationId: "Default",
        locationPath: exports.DEFAULT_TABLE_PERSISTENCE_PATH,
        maxConcurrency: 1
    }
];
exports.QUERY_RESULT_MAX_NUM = 1000;
exports.ValidAPIVersions = [
    "2024-05-04",
    "2024-02-04",
    "2023-11-03",
    "2023-08-03",
    "2023-01-03",
    "2022-11-02",
    "2021-12-02",
    "2021-10-04",
    "2021-08-06",
    "2021-06-08",
    "2021-04-10",
    "2021-02-12",
    "2020-12-06",
    "2020-10-02",
    "2020-08-04",
    "2020-06-12",
    "2020-04-08",
    "2020-02-10",
    "2019-12-12",
    "2019-10-10",
    "2019-07-07",
    "2019-02-02",
    "2018-11-09",
    "2018-03-28",
    "2017-11-09",
    "2017-07-29",
    "2017-04-17",
    "2016-05-31",
    "2015-12-11",
    "2015-07-08",
    "2015-04-05",
    "2015-02-21",
    "2014-02-14",
    "2013-08-15",
    "2012-02-12",
    "2011-08-18",
    "2009-09-19",
    "2009-07-17",
    "2009-04-14"
];
exports.TABLE_SERVICE_PERMISSION = "raud";
exports.SECONDARY_SUFFIX = "-secondary";
// Validate audience, accept following audience patterns
// https://storage.azure.com
// https://storage.azure.com/
// e406a681-f3d4-42a8-90b6-c2b029497af1
// https://*.table.core.windows.net
// https://*.table.core.windows.net/
// https://*.table.core.chinacloudapi.cn
// https://*.table.core.chinacloudapi.cn/
// https://*.table.core.usgovcloudapi.net
// https://*.table.core.usgovcloudapi.net/
// https://*.table.core.cloudapi.de
// https://*.table.core.cloudapi.de/
exports.VALID_TABLE_AUDIENCES = [
    /^https:\/\/storage\.azure\.com[\/]?$/,
    /^e406a681-f3d4-42a8-90b6-c2b029497af1$/,
    /^https:\/\/(.*)\.table\.core\.windows\.net[\/]?$/,
    /^https:\/\/(.*)\.table\.core\.chinacloudapi\.cn[\/]?$/,
    /^https:\/\/(.*)\.table\.core\.usgovcloudapi\.net[\/]?$/,
    /^https:\/\/(.*)\.table\.core\.cloudapi\.de[\/]?$/
];
