// azurite/dist/src/table/generated/artifacts/models.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Version16 = exports.Version15 = exports.Version14 = exports.Version13 = exports.Version12 = exports.Version11 = exports.Version10 = exports.Version9 = exports.Version8 = exports.Version7 = exports.Version6 = exports.Version5 = exports.Version4 = exports.Version3 = exports.Version2 = exports.Version1 = exports.DataServiceVersion9 = exports.DataServiceVersion8 = exports.DataServiceVersion7 = exports.DataServiceVersion6 = exports.DataServiceVersion5 = exports.DataServiceVersion4 = exports.DataServiceVersion3 = exports.DataServiceVersion2 = exports.DataServiceVersion1 = exports.DataServiceVersion = exports.Version = exports.ResponseFormat = exports.OdataMetadataFormat = exports.GeoReplicationStatusType = void 0;
/**
 * Defines values for GeoReplicationStatusType.
 * Possible values include: 'live', 'bootstrap', 'unavailable'
 * @readonly
 * @enum {string}
 */
var GeoReplicationStatusType;
(function (GeoReplicationStatusType) {
    GeoReplicationStatusType["Live"] = "live";
    GeoReplicationStatusType["Bootstrap"] = "bootstrap";
    GeoReplicationStatusType["Unavailable"] = "unavailable";
})(GeoReplicationStatusType || (exports.GeoReplicationStatusType = GeoReplicationStatusType = {}));
/**
 * Defines values for OdataMetadataFormat.
 * Possible values include: 'application/json;odata=nometadata',
 * 'application/json;odata=minimalmetadata', 'application/json;odata=fullmetadata'
 * @readonly
 * @enum {string}
 */
var OdataMetadataFormat;
(function (OdataMetadataFormat) {
    OdataMetadataFormat["Applicationjsonodatanometadata"] = "application/json;odata=nometadata";
    OdataMetadataFormat["Applicationjsonodataminimalmetadata"] = "application/json;odata=minimalmetadata";
    OdataMetadataFormat["Applicationjsonodatafullmetadata"] = "application/json;odata=fullmetadata";
})(OdataMetadataFormat || (exports.OdataMetadataFormat = OdataMetadataFormat = {}));
/**
 * Defines values for ResponseFormat.
 * Possible values include: 'return-no-content', 'return-content'
 * @readonly
 * @enum {string}
 */
var ResponseFormat;
(function (ResponseFormat) {
    ResponseFormat["ReturnNoContent"] = "return-no-content";
    ResponseFormat["ReturnContent"] = "return-content";
})(ResponseFormat || (exports.ResponseFormat = ResponseFormat = {}));
/**
 * Defines values for Version.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version;
(function (Version) {
    Version["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version || (exports.Version = Version = {}));
/**
 * Defines values for DataServiceVersion.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion;
(function (DataServiceVersion) {
    DataServiceVersion["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion || (exports.DataServiceVersion = DataServiceVersion = {}));
/**
 * Defines values for DataServiceVersion1.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion1;
(function (DataServiceVersion1) {
    DataServiceVersion1["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion1 || (exports.DataServiceVersion1 = DataServiceVersion1 = {}));
/**
 * Defines values for DataServiceVersion2.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion2;
(function (DataServiceVersion2) {
    DataServiceVersion2["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion2 || (exports.DataServiceVersion2 = DataServiceVersion2 = {}));
/**
 * Defines values for DataServiceVersion3.
 * Possible values include: '3.0'
 * Changing all table batch related logic to point to this version
 * becuase the other versions seem redundant and are complicating the code
 * unnecessarily
 * @readonly
 * @enum {string}
 */
var DataServiceVersion3;
(function (DataServiceVersion3) {
    DataServiceVersion3["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion3 || (exports.DataServiceVersion3 = DataServiceVersion3 = {}));
/**
 * Defines values for DataServiceVersion4.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion4;
(function (DataServiceVersion4) {
    DataServiceVersion4["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion4 || (exports.DataServiceVersion4 = DataServiceVersion4 = {}));
/**
 * Defines values for DataServiceVersion5.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion5;
(function (DataServiceVersion5) {
    DataServiceVersion5["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion5 || (exports.DataServiceVersion5 = DataServiceVersion5 = {}));
/**
 * Defines values for DataServiceVersion6.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion6;
(function (DataServiceVersion6) {
    DataServiceVersion6["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion6 || (exports.DataServiceVersion6 = DataServiceVersion6 = {}));
/**
 * Defines values for DataServiceVersion7.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion7;
(function (DataServiceVersion7) {
    DataServiceVersion7["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion7 || (exports.DataServiceVersion7 = DataServiceVersion7 = {}));
/**
 * Defines values for DataServiceVersion8.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion8;
(function (DataServiceVersion8) {
    DataServiceVersion8["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion8 || (exports.DataServiceVersion8 = DataServiceVersion8 = {}));
/**
 * Defines values for DataServiceVersion9.
 * Possible values include: '3.0'
 * @readonly
 * @enum {string}
 */
var DataServiceVersion9;
(function (DataServiceVersion9) {
    DataServiceVersion9["ThreeFullStopZero"] = "3.0";
})(DataServiceVersion9 || (exports.DataServiceVersion9 = DataServiceVersion9 = {}));
/**
 * Defines values for Version1.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version1;
(function (Version1) {
    Version1["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version1 || (exports.Version1 = Version1 = {}));
/**
 * Defines values for Version2.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version2;
(function (Version2) {
    Version2["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version2 || (exports.Version2 = Version2 = {}));
/**
 * Defines values for Version3.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version3;
(function (Version3) {
    Version3["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version3 || (exports.Version3 = Version3 = {}));
/**
 * Defines values for Version4.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version4;
(function (Version4) {
    Version4["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version4 || (exports.Version4 = Version4 = {}));
/**
 * Defines values for Version5.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version5;
(function (Version5) {
    Version5["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version5 || (exports.Version5 = Version5 = {}));
/**
 * Defines values for Version6.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version6;
(function (Version6) {
    Version6["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version6 || (exports.Version6 = Version6 = {}));
/**
 * Defines values for Version7.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version7;
(function (Version7) {
    Version7["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version7 || (exports.Version7 = Version7 = {}));
/**
 * Defines values for Version8.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version8;
(function (Version8) {
    Version8["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version8 || (exports.Version8 = Version8 = {}));
/**
 * Defines values for Version9.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version9;
(function (Version9) {
    Version9["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version9 || (exports.Version9 = Version9 = {}));
/**
 * Defines values for Version10.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version10;
(function (Version10) {
    Version10["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version10 || (exports.Version10 = Version10 = {}));
/**
 * Defines values for Version11.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version11;
(function (Version11) {
    Version11["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version11 || (exports.Version11 = Version11 = {}));
/**
 * Defines values for Version12.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version12;
(function (Version12) {
    Version12["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version12 || (exports.Version12 = Version12 = {}));
/**
 * Defines values for Version13.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version13;
(function (Version13) {
    Version13["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version13 || (exports.Version13 = Version13 = {}));
/**
 * Defines values for Version14.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version14;
(function (Version14) {
    Version14["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version14 || (exports.Version14 = Version14 = {}));
/**
 * Defines values for Version15.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version15;
(function (Version15) {
    Version15["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version15 || (exports.Version15 = Version15 = {}));
/**
 * Defines values for Version16.
 * Possible values include: '2019-02-02'
 * @readonly
 * @enum {string}
 */
var Version16;
(function (Version16) {
    Version16["TwoZeroOneNineHyphenMinusZeroTwoHyphenMinusZeroTwo"] = "2019-02-02";
})(Version16 || (exports.Version16 = Version16 = {}));
