// azurite/dist/src/table/generated/artifacts/operation.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operation = void 0;
var Operation;
(function (Operation) {
    Operation[Operation["Table_Query"] = 0] = "Table_Query";
    Operation[Operation["Table_Create"] = 1] = "Table_Create";
    Operation[Operation["Table_Batch"] = 2] = "Table_Batch";
    Operation[Operation["Table_Delete"] = 3] = "Table_Delete";
    Operation[Operation["Table_QueryEntities"] = 4] = "Table_QueryEntities";
    Operation[Operation["Table_QueryEntitiesWithPartitionAndRowKey"] = 5] = "Table_QueryEntitiesWithPartitionAndRowKey";
    Operation[Operation["Table_UpdateEntity"] = 6] = "Table_UpdateEntity";
    Operation[Operation["Table_MergeEntity"] = 7] = "Table_MergeEntity";
    Operation[Operation["Table_DeleteEntity"] = 8] = "Table_DeleteEntity";
    Operation[Operation["Table_MergeEntityWithMerge"] = 9] = "Table_MergeEntityWithMerge";
    Operation[Operation["Table_InsertEntity"] = 10] = "Table_InsertEntity";
    Operation[Operation["Table_GetAccessPolicy"] = 11] = "Table_GetAccessPolicy";
    Operation[Operation["Table_SetAccessPolicy"] = 12] = "Table_SetAccessPolicy";
    Operation[Operation["Service_SetProperties"] = 13] = "Service_SetProperties";
    Operation[Operation["Service_GetProperties"] = 14] = "Service_GetProperties";
    Operation[Operation["Service_GetStatistics"] = 15] = "Service_GetStatistics";
})(Operation || (exports.Operation = Operation = {}));
exports.default = Operation;
