// azurite/dist/src/blob/context/BlobStorageContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Context_1 = tslib_1.__importDefault(require("../generated/Context"));
class BlobStorageContext extends Context_1.default {
    getContainer() {
        return this.context.container;
    }
    get account() {
        return this.context.account;
    }
    set account(account) {
        this.context.account = account;
    }
    set isSecondary(isSecondary) {
        this.context.isSecondary = isSecondary;
    }
    get isSecondary() {
        return this.context.isSecondary;
    }
    get container() {
        return this.context.container;
    }
    set container(container) {
        this.context.container = container;
    }
    get blob() {
        return this.context.blob;
    }
    set blob(blob) {
        this.context.blob = blob;
    }
    get authenticationPath() {
        return this.context.authenticationPath;
    }
    set authenticationPath(path) {
        this.context.authenticationPath = path;
    }
    get xMsRequestID() {
        return this.contextId;
    }
    set xMsRequestID(xMsRequestID) {
        this.contextId = xMsRequestID;
    }
    get disableProductStyleUrl() {
        return this.context.disableProductStyleUrl;
    }
    set disableProductStyleUrl(disableProductStyleUrl) {
        this.context.disableProductStyleUrl = disableProductStyleUrl;
    }
    get loose() {
        return this.context.loose;
    }
    set loose(loose) {
        this.context.loose = loose;
    }
}
exports.default = BlobStorageContext;
