// azurite/dist/src/queue/context/QueueStorageContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Context_1 = tslib_1.__importDefault(require("../generated/Context"));
class QueueStorageContext extends Context_1.default {
    get account() {
        return this.context.account;
    }
    set account(account) {
        this.context.account = account;
    }
    set isSecondary(isSecondary) {
        this.context.isSecondary = isSecondary;
    }
    get isSecondary() {
        return this.context.isSecondary;
    }
    get queue() {
        return this.context.queue;
    }
    set queue(queue) {
        this.context.queue = queue;
    }
    get message() {
        return this.context.message;
    }
    set message(message) {
        this.context.message = message;
    }
    get messageId() {
        return this.context.messageId;
    }
    set messageId(messageId) {
        this.context.messageId = messageId;
    }
    get authenticationPath() {
        return this.context.authenticationPath;
    }
    set authenticationPath(path) {
        this.context.authenticationPath = path;
    }
    get xMsRequestID() {
        return this.contextID;
    }
    set xMsRequestID(xMsRequestID) {
        this.contextID = xMsRequestID;
    }
}
exports.default = QueueStorageContext;
