// azurite/dist/src/common/VSCNotification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_1 = require("vscode");
class VSCNotification {
    onStart(serverManager, session) {
        // Noop
    }
    onStartFail(serverManager, session, error) {
        vscode_1.window.showErrorMessage(`Start ${serverManager.name} error: ${error.message}`);
    }
    onStartSuccess(serverManager, session) {
        vscode_1.window.showInformationMessage(`${serverManager.name} successfully listens on ${serverManager
            .getServer()
            .getHttpServerAddress()}`);
    }
    onClean(serverManager, session) {
        // Noop
    }
    onCleanFail(serverManager, session, error) {
        vscode_1.window.showErrorMessage(`Clean ${serverManager.name} error: ${error.message}`);
    }
    onCleanSuccess(serverManager, session) {
        vscode_1.window.showInformationMessage(`Clean up ${serverManager.name} successfully.`);
    }
    onClose(serverManager, session) {
        // Noop
    }
    onCloseFail(serverManager, session, error) {
        vscode_1.window.showErrorMessage(`Close ${serverManager.name} error: ${error.message}`);
    }
    onCloseSuccess(serverManager, session) {
        vscode_1.window.showInformationMessage(`${serverManager.name} successfully closed.`);
    }
}
exports.default = VSCNotification;
