// azurite/dist/src/table/TableConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const ConfigurationBase_1 = tslib_1.__importDefault(require("../common/ConfigurationBase"));
const constants_1 = require("./utils/constants");
/**
 * Default configurations for default implementation of TableServer.
 *
 * As default implementation of TableServer class leverages LokiJS DB.
 * This configuration class also maintains configuration settings for LokiJS DB.
 *
 * When creating other server implementations, should also create a NEW
 * corresponding configuration class by extending ConfigurationBase.
 *
 * @export
 * @class Configuration
 */
class TableConfiguration extends ConfigurationBase_1.default {
    constructor(host = constants_1.DEFAULT_TABLE_SERVER_HOST_NAME, port = constants_1.DEFAULT_TABLE_LISTENING_PORT, metadataDBPath = constants_1.DEFAULT_TABLE_LOKI_DB_PATH, enableDebugLog = constants_1.DEFAULT_ENABLE_DEBUG_LOG, enableAccessLog = constants_1.DEFAULT_ENABLE_ACCESS_LOG, accessLogWriteStream, debugLogFilePath, loose = false, skipApiVersionCheck = false, cert = "", key = "", pwd = "", oauth, disableProductStyleUrl = false, isMemoryPersistence = false) {
        super(host, port, enableAccessLog, accessLogWriteStream, enableDebugLog, debugLogFilePath, loose, skipApiVersionCheck, cert, key, pwd, oauth, disableProductStyleUrl);
        this.metadataDBPath = metadataDBPath;
        this.isMemoryPersistence = isMemoryPersistence;
    }
}
exports.default = TableConfiguration;
