// azurite/dist/src/table/entity/EdmBoolean.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdmBoolean = void 0;
class EdmBoolean {
    static validate(value) {
        if (typeof value === "string") {
            // assuming that we do not need to use value.toLocaleLowerCase()
            if (value !== "true" && value !== "false") {
                throw TypeError(`Not a valid EdmBoolean string.`);
            }
            return value === "true";
        }
        else if (typeof value !== "boolean") {
            throw TypeError(`Not a valid EdmBoolean string.`);
        }
        return value;
    }
    constructor(value) {
        this.value = value;
        this.typedValue = EdmBoolean.validate(value);
    }
    toJsonPropertyValuePair(name) {
        return [name, this.typedValue];
    }
    toJsonPropertyValueString(name) {
        return `"${name}":${this.typedValue}`;
    }
    toJsonPropertyTypePair(_name, _annotationLevel, _isSystemProperty) {
        return;
    }
    toJsonPropertyTypeString(_name, _annotationLevel, _isSystemProperty) {
        return;
    }
}
exports.EdmBoolean = EdmBoolean;
