// azurite/dist/src/blob/lease/ContainerLeaseAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ContainerLeaseAdapter {
    constructor(container) {
        if (container.properties.leaseState === undefined) {
            throw RangeError(`ContainerLeaseAdapter:constructor() container leaseState cannot be undefined.`);
        }
        if (container.properties.leaseStatus === undefined) {
            throw RangeError(`ContainerLeaseAdapter:constructor() container leaseStatus cannot be undefined.`);
        }
        this.leaseId = container.leaseId;
        this.leaseState = container.properties.leaseState;
        this.leaseStatus = container.properties.leaseStatus;
        this.leaseDurationType = container.properties.leaseDuration;
        this.leaseDurationSeconds = container.leaseDurationSeconds;
        this.leaseExpireTime = container.leaseExpireTime;
        this.leaseBreakTime = container.leaseBreakTime;
    }
    toString() {
        return JSON.stringify({
            leaseId: this.leaseId,
            leaseState: this.leaseState,
            leaseStatus: this.leaseStatus,
            leaseDurationType: this.leaseDurationType,
            leaseDurationSeconds: this.leaseDurationSeconds,
            leaseExpireTime: this.leaseExpireTime,
            leaseBreakTime: this.leaseBreakTime
        });
    }
}
exports.default = ContainerLeaseAdapter;
