// azurite/dist/src/queue/generated/artifacts/operation.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operation = void 0;
var Operation;
(function (Operation) {
    Operation[Operation["Service_SetProperties"] = 0] = "Service_SetProperties";
    Operation[Operation["Service_GetProperties"] = 1] = "Service_GetProperties";
    Operation[Operation["Service_GetStatistics"] = 2] = "Service_GetStatistics";
    Operation[Operation["Service_ListQueuesSegment"] = 3] = "Service_ListQueuesSegment";
    Operation[Operation["Queue_Create"] = 4] = "Queue_Create";
    Operation[Operation["Queue_Delete"] = 5] = "Queue_Delete";
    Operation[Operation["Queue_GetProperties"] = 6] = "Queue_GetProperties";
    Operation[Operation["Queue_GetPropertiesWithHead"] = 7] = "Queue_GetPropertiesWithHead";
    Operation[Operation["Queue_SetMetadata"] = 8] = "Queue_SetMetadata";
    Operation[Operation["Queue_GetAccessPolicy"] = 9] = "Queue_GetAccessPolicy";
    Operation[Operation["Queue_GetAccessPolicyWithHead"] = 10] = "Queue_GetAccessPolicyWithHead";
    Operation[Operation["Queue_SetAccessPolicy"] = 11] = "Queue_SetAccessPolicy";
    Operation[Operation["Messages_Dequeue"] = 12] = "Messages_Dequeue";
    Operation[Operation["Messages_Clear"] = 13] = "Messages_Clear";
    Operation[Operation["Messages_Enqueue"] = 14] = "Messages_Enqueue";
    Operation[Operation["Messages_Peek"] = 15] = "Messages_Peek";
    Operation[Operation["MessageId_Update"] = 16] = "MessageId_Update";
    Operation[Operation["MessageId_Delete"] = 17] = "MessageId_Delete";
})(Operation || (exports.Operation = Operation = {}));
exports.default = Operation;
