// azurite/dist/src/queue/QueueConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const ConfigurationBase_1 = tslib_1.__importDefault(require("../common/ConfigurationBase"));
const constants_1 = require("./utils/constants");
/**
 * Default configurations for default implementation of QueueServer.
 *
 * As default implementation of QueueServer class leverages LokiJS DB.
 * This configuration class also maintains configuration settings for LokiJS DB.
 *
 * When creating other server implementations, should also create a NEW
 * corresponding configuration class by extending ConfigurationBase.
 *
 * @export
 * @class Configuration
 */
class QueueConfiguration extends ConfigurationBase_1.default {
    constructor(host = constants_1.DEFAULT_QUEUE_SERVER_HOST_NAME, port = constants_1.DEFAULT_QUEUE_LISTENING_PORT, metadataDBPath = constants_1.DEFAULT_QUEUE_LOKI_DB_PATH, extentDBPath = constants_1.DEFAULT_QUEUE_EXTENT_LOKI_DB_PATH, persistencePathArray = constants_1.DEFAULT_QUEUE_PERSISTENCE_ARRAY, enableAccessLog = constants_1.DEFAULT_ENABLE_ACCESS_LOG, accessLogWriteStream, enableDebugLog = constants_1.DEFAULT_ENABLE_DEBUG_LOG, debugLogFilePath, loose = false, skipApiVersionCheck = false, cert = "", key = "", pwd = "", oauth, disableProductStyleUrl = false, isMemoryPersistence = false, memoryStore) {
        super(host, port, enableAccessLog, accessLogWriteStream, enableDebugLog, debugLogFilePath, loose, skipApiVersionCheck, cert, key, pwd, oauth, disableProductStyleUrl);
        this.metadataDBPath = metadataDBPath;
        this.extentDBPath = extentDBPath;
        this.persistencePathArray = persistencePathArray;
        this.isMemoryPersistence = isMemoryPersistence;
        this.memoryStore = memoryStore;
    }
}
exports.default = QueueConfiguration;
