// azurite/dist/src/blob/generated/artifacts/operation.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operation = void 0;
var Operation;
(function (Operation) {
    Operation[Operation["Service_SetProperties"] = 0] = "Service_SetProperties";
    Operation[Operation["Service_GetProperties"] = 1] = "Service_GetProperties";
    Operation[Operation["Service_GetStatistics"] = 2] = "Service_GetStatistics";
    Operation[Operation["Service_ListContainersSegment"] = 3] = "Service_ListContainersSegment";
    Operation[Operation["Service_GetUserDelegationKey"] = 4] = "Service_GetUserDelegationKey";
    Operation[Operation["Service_GetAccountInfo"] = 5] = "Service_GetAccountInfo";
    Operation[Operation["Service_GetAccountInfoWithHead"] = 6] = "Service_GetAccountInfoWithHead";
    Operation[Operation["Service_SubmitBatch"] = 7] = "Service_SubmitBatch";
    Operation[Operation["Service_FilterBlobs"] = 8] = "Service_FilterBlobs";
    Operation[Operation["Container_Create"] = 9] = "Container_Create";
    Operation[Operation["Container_GetProperties"] = 10] = "Container_GetProperties";
    Operation[Operation["Container_GetPropertiesWithHead"] = 11] = "Container_GetPropertiesWithHead";
    Operation[Operation["Container_Delete"] = 12] = "Container_Delete";
    Operation[Operation["Container_SetMetadata"] = 13] = "Container_SetMetadata";
    Operation[Operation["Container_GetAccessPolicy"] = 14] = "Container_GetAccessPolicy";
    Operation[Operation["Container_SetAccessPolicy"] = 15] = "Container_SetAccessPolicy";
    Operation[Operation["Container_Restore"] = 16] = "Container_Restore";
    Operation[Operation["Container_SubmitBatch"] = 17] = "Container_SubmitBatch";
    Operation[Operation["Container_FilterBlobs"] = 18] = "Container_FilterBlobs";
    Operation[Operation["Container_AcquireLease"] = 19] = "Container_AcquireLease";
    Operation[Operation["Container_ReleaseLease"] = 20] = "Container_ReleaseLease";
    Operation[Operation["Container_RenewLease"] = 21] = "Container_RenewLease";
    Operation[Operation["Container_BreakLease"] = 22] = "Container_BreakLease";
    Operation[Operation["Container_ChangeLease"] = 23] = "Container_ChangeLease";
    Operation[Operation["Container_ListBlobFlatSegment"] = 24] = "Container_ListBlobFlatSegment";
    Operation[Operation["Container_ListBlobHierarchySegment"] = 25] = "Container_ListBlobHierarchySegment";
    Operation[Operation["Container_GetAccountInfo"] = 26] = "Container_GetAccountInfo";
    Operation[Operation["Container_GetAccountInfoWithHead"] = 27] = "Container_GetAccountInfoWithHead";
    Operation[Operation["Blob_Download"] = 28] = "Blob_Download";
    Operation[Operation["Blob_GetProperties"] = 29] = "Blob_GetProperties";
    Operation[Operation["Blob_Delete"] = 30] = "Blob_Delete";
    Operation[Operation["Blob_Undelete"] = 31] = "Blob_Undelete";
    Operation[Operation["Blob_SetExpiry"] = 32] = "Blob_SetExpiry";
    Operation[Operation["Blob_SetHTTPHeaders"] = 33] = "Blob_SetHTTPHeaders";
    Operation[Operation["Blob_SetImmutabilityPolicy"] = 34] = "Blob_SetImmutabilityPolicy";
    Operation[Operation["Blob_DeleteImmutabilityPolicy"] = 35] = "Blob_DeleteImmutabilityPolicy";
    Operation[Operation["Blob_SetLegalHold"] = 36] = "Blob_SetLegalHold";
    Operation[Operation["Blob_SetMetadata"] = 37] = "Blob_SetMetadata";
    Operation[Operation["Blob_AcquireLease"] = 38] = "Blob_AcquireLease";
    Operation[Operation["Blob_ReleaseLease"] = 39] = "Blob_ReleaseLease";
    Operation[Operation["Blob_RenewLease"] = 40] = "Blob_RenewLease";
    Operation[Operation["Blob_ChangeLease"] = 41] = "Blob_ChangeLease";
    Operation[Operation["Blob_BreakLease"] = 42] = "Blob_BreakLease";
    Operation[Operation["Blob_CreateSnapshot"] = 43] = "Blob_CreateSnapshot";
    Operation[Operation["Blob_StartCopyFromURL"] = 44] = "Blob_StartCopyFromURL";
    Operation[Operation["Blob_CopyFromURL"] = 45] = "Blob_CopyFromURL";
    Operation[Operation["Blob_AbortCopyFromURL"] = 46] = "Blob_AbortCopyFromURL";
    Operation[Operation["Blob_SetTier"] = 47] = "Blob_SetTier";
    Operation[Operation["Blob_GetAccountInfo"] = 48] = "Blob_GetAccountInfo";
    Operation[Operation["Blob_GetAccountInfoWithHead"] = 49] = "Blob_GetAccountInfoWithHead";
    Operation[Operation["Blob_Query"] = 50] = "Blob_Query";
    Operation[Operation["Blob_GetTags"] = 51] = "Blob_GetTags";
    Operation[Operation["Blob_SetTags"] = 52] = "Blob_SetTags";
    Operation[Operation["PageBlob_Create"] = 53] = "PageBlob_Create";
    Operation[Operation["PageBlob_UploadPages"] = 54] = "PageBlob_UploadPages";
    Operation[Operation["PageBlob_ClearPages"] = 55] = "PageBlob_ClearPages";
    Operation[Operation["PageBlob_UploadPagesFromURL"] = 56] = "PageBlob_UploadPagesFromURL";
    Operation[Operation["PageBlob_GetPageRanges"] = 57] = "PageBlob_GetPageRanges";
    Operation[Operation["PageBlob_GetPageRangesDiff"] = 58] = "PageBlob_GetPageRangesDiff";
    Operation[Operation["PageBlob_Resize"] = 59] = "PageBlob_Resize";
    Operation[Operation["PageBlob_UpdateSequenceNumber"] = 60] = "PageBlob_UpdateSequenceNumber";
    Operation[Operation["PageBlob_CopyIncremental"] = 61] = "PageBlob_CopyIncremental";
    Operation[Operation["AppendBlob_Create"] = 62] = "AppendBlob_Create";
    Operation[Operation["AppendBlob_AppendBlock"] = 63] = "AppendBlob_AppendBlock";
    Operation[Operation["AppendBlob_AppendBlockFromUrl"] = 64] = "AppendBlob_AppendBlockFromUrl";
    Operation[Operation["AppendBlob_Seal"] = 65] = "AppendBlob_Seal";
    Operation[Operation["BlockBlob_Upload"] = 66] = "BlockBlob_Upload";
    Operation[Operation["BlockBlob_PutBlobFromUrl"] = 67] = "BlockBlob_PutBlobFromUrl";
    Operation[Operation["BlockBlob_StageBlock"] = 68] = "BlockBlob_StageBlock";
    Operation[Operation["BlockBlob_StageBlockFromURL"] = 69] = "BlockBlob_StageBlockFromURL";
    Operation[Operation["BlockBlob_CommitBlockList"] = 70] = "BlockBlob_CommitBlockList";
    Operation[Operation["BlockBlob_GetBlockList"] = 71] = "BlockBlob_GetBlockList";
})(Operation || (exports.Operation = Operation = {}));
exports.default = Operation;
